Compiler support to accelerate nd_range parallel_for on CPU

fodinabor opened the pull request:
This adds explicit compiler support for the nd_range parallel_for paradigm to speed it up by up to multiple orders of magnitude compared to the current Boost.Fiber-based implementation.

Exemplary performance comparison of https://github.com/UoB-HPC/sycl_dgemm
Shows fibers, the compiler supported CPU version `cbs_man`, the hierarchical implementation as well as an equivalent OpenCL implementation using differing POCL modes on 4 different machines (all from Isambard, big thanks to @simonmcs for letting us test there!).

![sycl_dgemm csv](https://user-images.githubusercontent.com/5982050/150011965-48426602-fa13-483e-af91-711810e9459e.png)

Same game for SYCL-Bench's reduction benchmark:
![sycl-bench-red csv](https://user-images.githubusercontent.com/5982050/150012642-b81dd9f4-7019-4230-802d-fe4e871c46e6.png)

Both examples show that there's a real benefit to using this instead of the fiber nd_range variant! Though, there's still some room for improvement to fully match the hierarchical performance.

Copying the high-level documentation:  
>While hipSYCL usually strives to reuse as much compiler infrastructure as possible, the nd_range parallel_for is not efficiently implementable without explicit compiler support.
>
>Therefore, the hipSYCL Clang plugin implements a set of passes to perform deep loop fission on nd_range parallel_for kernels that contain barriers. The continuation-based synchronization approach is employed to achieve good performance and  functional correctness (_Karrenberg, Ralf, and Sebastian Hack. "Improving performance of OpenCL on CPUs." International Conference on Compiler Construction. Springer, Berlin, Heidelberg, 2012. [https://link.springer.com/content/pdf/10.1007/978-3-642-28652-0_1.pdf](https://link.springer.com/content/pdf/10.1007/978-3-642-28652-0_1.pdf)_). A deep dive into how the implementation works and why this approach was chosen can be found in my [master thesis](https://joameyer.de/hipsycl/Thesis_JoachimMeyer.pdf).
>
>As Clang plugins are only officially supported on Linux, this functionality is currently supported by hipSYCL on Linux only as well.
By default, this compiler support is used, as long as Clang is used as host compiler as well. It can be explicitly disabled, by setting `-DWITH_ACCELERATED_CPU=OFF` at hipSYCL build time. If hipSYCL was built with the compiler support enabled, it can be explicitly en-/disabled by adding `--hipsycl-use-accelerated-cpu`/`--hipsycl-use-accelerated-cpu=false` to the syclcc command line. When using CMake, set `-DHIPSYCL_USE_ACCELERATED_CPU=` to override the default behavior.

This currently is tested with LLVM 11-13 in the CI but is also compatible with LLVM 14 at the time of writing.
I would claim this feature still is experimental, but encourage everyone to test it out and would love any feedback!
If someone uses non-XCode Clang on Mac, it should be possible to enable the Clang plugin there as well, but I can't test it, so I didn't do it...

fodinabor committed:
Add basic infrastructure for nd_range barrier elimination.

This is done for the CPU backend with compiler support and therefore in the clang plugin.
In this first step,
- the OpenMP kernel launcher is adapted, to directly use OpenMP for nd_range kernels
- the plugin registers two new passes:
  * SplitterAnnotationAnalysisLegacy which builds up a list of functions annotated with "hipsycl_splitter".
  * LoopSplitAtBarrierPassLegacy which is targeted to do the actual splitting. Currently it contains preparatory code only: checking of the context, searching of splitter functions (barriers) and inlining of call trees, containing splitters.

fodinabor committed:
Add LLVM style .clang-format files to the compiler sub-folders.

fodinabor committed:
Transitively add nothrow attribute to OpenMP kernel functions.

As of the SYCL spec, functions in a kernel must not throw. Therefore it is safe to mark those as nothrow (aka noexcept).
This is done in the clang frontend plugin in a similar way as the kernel functions are marked with CUDAHost and CUDADevice attributes.
This change massively simplifies the CFG that has to be adapted when performing the loop splitting as there are no invoke instructions with their respective landing pads.

fodinabor committed:
Fix build with LLVM 10.

fodinabor committed:
Implement basic loop splitting for barriers directly in the work-item loop.

fodinabor committed:
First (unfinished) attempts to use have memory dependencies analyzed to enable recalculation of local variables.

As a work item lives inside a loop, if we split the loop, all private state of a work item (aka loop iteration) from the loop before, is lost.
Therefore a dependency hull is generated, that contains all Instructions, that interact with a Value, that depends on the loop iteration variable. Some filtering has to be applied, as e.g. we may not overwrite global memory again. This will have to be improved.

Mostly a backup commit, so I can rebase.

fodinabor committed:
Make syclcc-clang use clang + plugin for omp_invocation.

fodinabor committed:
Move clang::annotate in front of __device__ to fix CUDA build.

fodinabor committed:
Make Loop passes compatible with new LLVM PM.

There seems to be a dependency problem, though, as SplitterAnnotationAnalysis is not available in the Loop Pass. Maybe Loop passes may not depend on Module analyses or any other registration issue?

fodinabor committed:
[LoopSplit] Fix loop splitting, when latch == body.

In this case the loop induction variable (that we always have), would be removed from all but the last splits. Thus the first loops would never terminate.
As a solution to this, a latch is split of and the induction variable is moved into the latch.

fodinabor committed:
[LoopSplit] Refactor latch maker into MakeLatch function.

fodinabor committed:
[LoopSplit] Use eraseFromParent instead of removeFromParent.

fodinabor committed:
[LoopSplit] Use LLVM naming conventions.

fodinabor committed:
[LoopSplit] Add omp simd collapse to local loops for multi dimensional parallel_for_ndrange.

As this merges the two/three loops, the problem now is basically equivalent to the 1D case.

fodinabor committed:
[LoopSplit] Start moving towards arrayification instead of copying all depending instructions.

For each cross barrier dependency, we earlier had to copy all instructions to the new loop, which came with a bunch of difficulties, as having to determine which instructions were not overwritten yet, ..
Instead, now we keep state on stack. Thus, for each cross barrier dependency we create a 1024 elements (alloca) stack array, which we use to store the values for each iteration, which are needed again after the barrier.
This removes the need to recalculate everything, as well as the need to filter overwritten values and in theory makes handling branches much easier.

fodinabor committed:
[LoopSplit] Always simplify latch and correctly exclude exit block.

fodinabor committed:
[LoopSplit] Add reduction parameters to nd_range work items.

fodinabor committed:
[LoopSplit] Highly experimental support for barriers in conditionals.

This first performs an analysis to find all the conditionals in the work item. Then given this information, and whether a conditional contains a barrier, this conditional is cloned and rewired, s.t. only the conditional path with the current barrier in it stays in the first loop (as we take care of one barrier at a time). The part leading to the conditional stays in the first loop. The other paths are moved into the next loop. All conditionals are cloned, but only the comparison and the branch operator.

The value dependency analysis is now moved after rewriting the loop, as we then know all blocks which are in the new and the old loops. The initial alloca arrayification stays at the beginning of the process.

DT and Loop analysis are not taken care of too much during the process right now, but using the updateDTAndLi they are still preserved.

fodinabor committed:
[LoopSplit] Experimental support for loops inside work items.

Only a single loop depth is supported for now. Also the type of loops supported is pretty limited as well. E.g. the loop iteration variables may not depend on the work item id. Non uniform (e.g. including break) loops are entirely untested as well.

fodinabor committed:
[LoopSplit] Add vectorization MD to latch terminator.

This should tell the vectorizer to vectorize the work-item loops.

fodinabor committed:
[LoopSplit] Actually enable vectorization.

This adds `#pragma omp simd` to the work item loop, to aid some array bounds check issues. And annotates all added alloca stuff with `llvm.access.group` and adds the access groups to the `llvm.loop.parallel_accesses`.

fodinabor committed:
[LoopSplit] Support more inner loops.

This adds highly experimental support for inner loops with non-constant initial loop variables and non-uniform variables. I.e. it is possible now, to have differing actual loop indices between work items (as long as the amount of iterations is the same between all work items).

This adds a bunch of todos as this is probably rather brittle and needs refactoring.

fodinabor committed:
[LoopSplit] Add LLVM lit config to improve testing of the plugin.

It is now possible to use the ease of LLVM lit to test many small scenarios to check we don't break anything.
This adds some variations of a reduction testing various aspects that should work so far as well as some barriers in conditionals.

Also move instruction simplification to the end of the pass (which we probably could leave to later passes then) to allow reduce_for_inverse_barrier.cpp.

fodinabor committed:
[LoopSplit] Support SYCL 2020's sycl::group_barrier(g) as splitter.

fodinabor committed:
[LoopSplit] Add HIPSYCL_DEBUG_EXECUTE_* to be able to control cfg generation and more at runtime.

fodinabor committed:
[LoopSplit] Allow additional non-induction values in inner loops like accumulators.

Also allows multiple induction variables.

fodinabor committed:
[LoopSplit] Fix inner loops with multiple barriers.

fodinabor committed:
[LoopSplit] Merge llvm.access.group alloca load MD if necessary.

fodinabor committed:
[LoopSplit] Unit tests build and run successfully!

Except for scoped parallel reduction.. but that also fails when not setting HIPSYCL_NO_FIBERS and the loops splitter should not be at work at all.

For this to work, the condition finding had to be slightly generalized and constant initialized accumulators had to be fixed.
Also the local_memory_ptr was uninitialized.

fodinabor committed:
[LoopSplit] Don't require -fno-discard-value-names anymore.

Instead of using a blacklist of value names to filter out allocas that are not work-item private, perform an usage analysis to identify allocas that are exclusively used in the the work-item loop. Thus we don't need to add that flag anymore.

fodinabor committed:
[LoopSplit] Get InnerLoop again after first split to fix BabelStream's dot kernel.

fodinabor committed:
[LoopSplit] Arrayify dependencies between inner loop body blocks and loop latch.

This allows SYCL-Bench's scalar prod to compile successfully.

fodinabor committed:
[LoopSplit] Allow LoopSplitter to build with O0.

Some passes the loop splitter relies on, are not run with O0 by default.
So the transformations are executed manually (again).

fodinabor committed:
[LoopSplit] Rework handling of inner loop induction variables.

Instead of the brittle two parts used before, now the inner latch is just cleared completely and by this added to the last work-item loop with the depended upon values being stored to the corresponding alloca and loaded again in the inner latch to be available for the header's PHINodes.

Also document the process of the inner loop split a bit and slightly clean up the code.

fodinabor committed:
[LoopSplit] Add now passing regression test with induction variable update not at the end of loop.

fodinabor committed:
Merge remote-tracking branch 'origin/develop' into feature/cpu_loop_splitting

fodinabor committed:
[LoopSplit] Update nd range kernel launcher and kernel detection.

After merging current develop, the kernel launcher had to be updated to reflect the changes done for the fiber path.
This resulted in the omp parallel region to be no longer outlined in a separate function. Thus the mechanism to find kernels to split had to be adapted: the `iterate_nd_range_omp` is `clang::annotate`d with `hipsycl_nd_kernel` which can be found in the same way as the `hipsycl_splitter` annotation.
As this is also used in the frontend for the noexcept sprinkling, the dependency on using OpenMP is highly reduced.

fodinabor committed:
Merge current origin/develop.

fodinabor committed:
Add workaround for failures with multiple reductions in nd parallel for.

fodinabor committed:
[LoopSplit] Enable nested inner loops.

This allows the use of multiple nested loops inside of the work items with loop splitting.
This is done by first collecting all inner loops that have to be split (aka all non work item loops) and then, from the outermost to the innermost loop, repeatedly applying the loop inversing mechanism as used before. This way, the inner loop being split, is always directly inside a valid work item loop.

fodinabor committed:
[LoopSplit] Only run O0 simplify when actually in O0.

fodinabor committed:
[LoopSplit] Insert llvm.lifetime.end for arrayified allocas.

Also mark every memory access work-item parallel.

fodinabor committed:
[LoopSplit] Attempts at making kernel code better debuggable.

Removing debug locations from the work item headers, so one doesn't jump back and forth.
Also cloning dbg value instructions when loading from arrayified alloca to inform the debugger what the value corresponds to.

illuhad committed:
Add static property list

illuhad committed:
Expose static property list to kernel launchers

illuhad committed:
Implement checking for valid work group sizes using reqd_work_group_size and add test case

illuhad committed:
Merge branch 'develop' into feature/static-kernel-properties

fodinabor committed:
[LoopSplit] Experimentally pool the loads into a load block.

fodinabor committed:
[LoopSplit] Fix getting variable values in debugger.

The this pointer of the functor is now copied to the load block as well, to allow depending values to be resolved.
In the debug values of the inner loop induction values were not handled correctly.

Also moveInnerIndVarToWorkItemLoop is required for inner loops.

Some tests are still failing due to missing array loads in the inner loop's header.

fodinabor committed:
[LoopSplit] Fix issue with accumulators outliving inner loops.

This surfaced in the PRK dgemm kernel with the results for the non 0st work item being wrong.

fodinabor committed:
[LoopSplit] Mark sycl-bench_scalar_prod as flaky, as it contains an invalid WAR dependency between workgroups.

fodinabor committed:
[LoopSplit] Fix handling of non constant initialized inner loop induction variables.

They now always are moved to their own alloca to avoid overriding the original value that might be needed both inside the loop and after the loop.

fodinabor committed:
[LoopSplit] Move inner ind var load to load pool.

fodinabor committed:
Merge branch 'feature/cpu_loop_splitting' into feature/experimental_load_pooling

fodinabor committed:
[LoopSplit] Add failing tests for in cond and do while.

fodinabor committed:
[LoopSplit] Move inner non-ind var loads to load pool.

fodinabor committed:
[LoopSplit] Slightly improve debug experience again.

fodinabor committed:
Merge remote-tracking branch 'origin/develop' into feature/cpu_loop_splitting

fodinabor committed:
[LoopSplit] Code clean up: split passes in separate files.

Also split out LoopSplitterInlining pass, that inlines the call trees leading to a barrier.

fodinabor committed:
[LoopSplit] Fix running with the new PM.

This requires getting the function / module analysis from the parent analysis managers.

fodinabor committed:
[LoopSplit] Add Kernel Flattening as extra pass.

This allows to rather easily enable and disable the kernel flattening.

fodinabor committed:
[LoopSplit] Make marking all memory accesses as parallel a distinct pass.

This removes the need to pass the MDAccessGroup all along, thus cleaning up the LoopSplitting and also runs after the kernel flattening pass, thus being more thorough.

fodinabor committed:
[LoopSplit][NFC] Many small NFC changes hoping to improve read- and maintainability.

fodinabor committed:
[LoopSplit] Move marking loops vectorize.enable to mark parallel pass.

fodinabor committed:
[LoopSplit] Make for_in_cond test at least pass frontend.

fodinabor committed:
[LoopSplit] Make Linux CI almost useful for loop splitting.

fodinabor committed:
[LoopSplit] Add missing prefixes for opt levels.

fodinabor committed:
[LoopSplit] Add missing include Clang 12.

fodinabor committed:
[LoopSplit] Add missing parameter for simplifyCFG and Clang 12.

fodinabor committed:
[LoopSplit] Add missing includes in LoopSplitter.

fodinabor committed:
[LoopSplit] Try to make Clang 12's simplifyCFG happy.

fodinabor committed:
Make HIPSYCL_DEBUG_* use llvm::outs() in the Clang Plugin.

fodinabor committed:
[LoopSplit] Move marking the initial work item loop into a separate pass.

And rename MarkLoopsParallelPass to LoopsParallelMarkerPass for a more similar naming scheme.

fodinabor committed:
[LoopSplit] Plugin compiles with LLVM 12 now.

fodinabor committed:
[LoopSplit] Fix up simplifyO0 with Use new PM in CI.

fodinabor committed:
[LoopSplit] Make LoopSplitterInlining pass a function pass.

Also use HIPSYCL_DEBUG_* directly to also handle the output.

fodinabor committed:
[LoopSplit] Enable CI with loop splitting.

Currently the LIT tests don't run with LLVM 12, due to some issue with the legacy pass manager. The CPU tests are only run with LLVM 12, as the reductions fail anyways, but this way LIT and CPU tests are run by some version.

fodinabor committed:
[LoopSplit] Import VariableUniformityAnalysis and BarrierTailReplication from POCL.

This is WIP, piece by piece I aim to import POCL's kernel compiler and make it compatible with our infrastructure / use-case.

One major difference between our usecase and POCL's that has to be kept in mind, is that POCL works on a single kernel function most of the time and only in the end generates the work item loops around it. BTR seems to be fine with this, though.

fodinabor committed:
[LoopSplit] Import PHIsToAllocas pass.

But restrict to PHIs in the work-item loops.

fodinabor committed:
[LoopSplit] Import IsolateRegions pass.

fodinabor committed:
[LoopSplit] Add required loop barriers and remove barrier calls passes.

The first adds barriers to inner loops (b-loop) to enable correct splitting later, the latter removes all barrier calls.

fodinabor committed:
[LoopSplit] Add barrier canonicalization pass.

fodinabor committed:
[LoopSplit] Run ReqdLoopBarriers only on loops inside the work item loop.

fodinabor committed:
[LoopSplit] Run IsolateRegions only on regions that are inside the work item loop.

fodinabor committed:
[LoopSplit] Enable b-loop-less loops with POCL work item handling.

fodinabor committed:
[LoopSplit] Enable most regression tests with the POCL approach.

This is enabled by using a do-while like approach for the work-item loops, if the first work item was not peeled. Otherwise the for loop style is kept.
This also enables the pocl pipeline with the legacy pass manager.
Still failing are the conds_in_for, reduce_while_early_update and for_ind_cond regression tests.

fodinabor committed:
[LoopSplit] Fix issues with the legacy PM.

Change the used extension point to EarlyAsPossible to more closely match the EP used with the new PM.
The only non-passing regression test now is: conds_in_for.

fodinabor committed:
[LoopSplit] Fix conds_in_for test.

After BTR we need to re-add barriers in the single (due to simplify pass) latch for the parallel region formation to work correctly. This is done in CanonicalizeBarriers.

fodinabor committed:
[LoopSplit] Stop running transformations on empty kernels + update SAA with intrinsic.

fodinabor committed:
[LoopSplit] Add -fno-discard-value-names to compiler invocation.

The POCL approach relies at some points relies on value names. I'd prefer to refactor this away but as a work-around syclcc just instructs clang to not discard the value names.

fodinabor committed:
[LoopSplit] Enable selecting the compilation pipeline with HIPSYCL_PIPELINE environment variable.

To enable simpler comparison of the (to be) implemented loop splitting approaches, the approaches can be selected by setting the environment variable HIPSYCL_PIPELINE while compiling.

fodinabor committed:
[LoopSplit] Rename while-switch pipeline to continuation based synchronization.

This will be based on the approach from "Improving Performance of OpenCL on CPUs", Karrenberg and Hack 2012.

fodinabor committed:
[LoopSplit] First steps towards continuation based synchronization.

- Reuse CanonicalizeBarriers and RemoveBarriers passes,
- implement DFS to identify sub-CFGs in kernel,
- replicate sub-CFGs and add separate new WILoop around it.

Next up will be saving and restoring variables which live out the context and adding the while-switch structure around.

fodinabor committed:
[LoopSplit] Add save and restore of multi-cfg variables for CBS.

Mostly re-uses code and ideas from the org implementation: arrayify all allocas before any transformation, after creating the SubCFG view, arrayify all multi-cfg values and finally when replicating the sub cfgs, create a load block with all required loads immediately after the cloned WI-Header.

fodinabor committed:
[LoopSplit] Build while-switch around sub-cfgs.

This mainly replaces the original WILoop header with a block that just includes a switch branch on the LastBarrierID value mapping to the sub-cfgs.
By doing removeUnreachableBlocks, the old WILoop is completely wiped then.
Due to some context storing issues, some changes to save and restore have been necessary + adding of a last step that fixes up non-dominating dependencies in a sub-cfg.

The compiled program runs successful in debug, but has some stack access issues. This will be handled in a later commit.

fodinabor committed:
[LoopSplit] Start lifetime and initialize LastBarrierIdStorage.

(ASAN still is not happy, might have to initialize the full allocas at once?)

fodinabor committed:
[LoopSplit] Use remapped inst->alloca map in fixSingleSubCfgValues.

This fixes the last failing test: reduce_while_early_update. Thus all regression tests pass now with CBS!

fodinabor committed:
[LoopSplit] Replicate GEP in single cfg instead of loading.

This allows all unit tests to compile.

fodinabor committed:
[LoopSplit] Purge lifetime intrinsics.

This is to make ASAN assume the lifetime correctly instead of only marking the first iteration's elements as alive and filling the others' with poison.

fodinabor committed:
[LoopSplit] Possible performance improvements.

Add alignment to loop state allocas, fix LoopsParallelMarker running on the correct loop and re-run the same right before vectorization as e.g. SROA does not propagate llvm.access.group properly..

fodinabor committed:
[LoopSplit] Make plugin compile with LLVM 13.

fodinabor committed:
[LoopSplit] Revert moving nd_item and move alloca arrayifing after SubCFG formation.

fodinabor committed:
[LoopSplit] Make custom alloca arrayification for CBS.

This eliminates unnecessary arrayification for allocas that are only used in one subcfg. For this a hull of all users that may read/write or work on a pointer to the alloca is created and checked whether it's in a single cfg.
Todo: construct a test case where this might be the case for two subcfgs but actually having a dependency..

fodinabor committed:
Merge branch 'develop' into feature/cpu_loop_splitting

fodinabor committed:
[LoopSplit] Only open cfg in parallel marker if verbose.

fodinabor committed:
Merge remote-tracking branch 'origin/develop' into feature/cpu_loop_splitting

fodinabor committed:
[LoopSplit][POCL] Copy loop MD to old header in non-peel case.

If not peeling, the first iteration still is unconditionally executed -> the header is moved to the end of the loop -> it acts as latch. To get correct detection of work-item loops and so on, we must copy the llvm.loop md over to the header's terminator.

fodinabor committed:
[LoopSplit][CBS] Apply do-while optimization from POCL.

As it doesn't make any sense to invoke a kernel with no work-items we can assume to have at least a single work-item, thus have the loop condition check only at the very end of the wi-loop.

fodinabor committed:
[LoopSplit] Fix KernelFlattening preserved analyses.

Clang 13 actually asserted that the cfg was changed although CFG analysis was marked as preserved.

fodinabor committed:
[LoopSplit][CBS] Single-SubCFG values cannot blindly only be loaded in the load block.

If the value in question is e.g. used as induction variable in an inner loop, it is not sufficient to load it once per WI, but instead is has to be reloaded at every loop iteration -> load has to happen immediately before the first use.

Also add test for the for{if} edge case that is only supported by csb.

fodinabor committed:
[LoopSplit][CBS] Incorporate SROA into CBS pipeline.

This leads to muuch cleaner generated code -> less stores of nd_item bs -> vectorization.
Legacy PM not functional yet.

fodinabor committed:
[LoopSplit] Mitigate some issues with other pipelines than CBS.

E.g. the latch splitting made POCL pipeline pretty unhappy, so only do that if actually required. Also it seems the POCL pipeline doesn't really like new PM + O0, add a warning / error.

fodinabor committed:
[LoopSplit] Uncomment latch simplify.

fodinabor committed:
[LoopSplit] Small pass cleanups.

fodinabor committed:
[LoopSplit] Fix marking predeclared kernel functions noexcept.

The CCS does not create the full hull of called functions if the functions are predeclared, as they might not yet be defined when VisitFunctionDecl is called. Thus, the actual noexcept marking has to be done when the translation unit is complete, aka in `HandleTranslationUnit`.

Also add the correct tag for the inlining log messages and replace barrier functions with __hipsycl_barrier for simplicity.

fodinabor committed:
[LoopSplit] Don't perform IPSCCP and SROA on legacy PM in CBS.

This is due to the EP used being a FPM EP -> no Module passes (IPSCCP) are allowed..

fodinabor committed:
Enable CI for compilation tests

fodinabor committed:
[LoopSplit] Backport compiler fix from POCL.

fodinabor committed:
[CI] Only ignore sycl_tests failures

fodinabor committed:
[LoopSplit] Add print to SAA.

fodinabor committed:
[LoopSplit][CBS] Split after barriers, if cond branch.

The CBS pipeline just deletes barrier blocks. This is an issue if the branch was conditional. To fix this, the barrier canonicalization splits after barriers, if the branch is conditional.

fodinabor committed:
[LoopSplit] Move barrier up in the broken reduction test..

fodinabor committed:
[LoopSplit] Add regression test for a cond between two barriers.

fodinabor committed:
[LoopSplit] Allow per-arrayification number of elements.

Should've been used in conjunction with uniformity analysis, but had to scrap that..

fodinabor committed:
Disable Mac and Win CI.

fodinabor committed:
Merge remote-tracking branch 'origin/feature/static-kernel-properties' into feature/experimental-reqd-wg-size

fodinabor committed:
[LoopSplit] Use reqd_work_group_size attribute.

Using sycl::attribute<sycl::reqd_work_group_size<..>> now provides a means to get constant bounds with OpenMP.
Further optimizations might be to get the value from the mangled kernel name in the compiler and use it there to optimize the alloca size (and remove the code duplication in the kernel launcher).

fodinabor committed:
[LoopSplit] Parse reqd_work_group_size from kernel name.

With this, the context array size is now either the number of elements in reqd_work_group_size or the default of 1024.

fodinabor committed:
[LoopSplit] Cleanup output.

fodinabor committed:
Ensure unique kernel names with attributes in HIP/CUDA.

Adding the property list to the kernel name ensures that for e.g. every unique reqd_work_group_size, a unique kernel is generated. With this in place, compiler hints may be generated from the property lists without invalidly mixing kernels and attributes.

(Note: SPIR-V uses effective_type in conjunction with __builtin_unique_stable_name, which will have to be tested and given its removal and refactoring in Clang > 11, will have to be looked at anyways..)

fodinabor committed:
[LoopSplit] Add todo regarding adding regex for MS mangling.

fodinabor committed:
[LoopSplit] Remove flaky scalar prod regression test.

fodinabor committed:
Merge remote-tracking branch 'origin/develop' into feature/cpu_loop_splitting

# Conflicts:
#	include/hipSYCL/compiler/Frontend.hpp
#	include/hipSYCL/compiler/IR.hpp
#	src/compiler/CMakeLists.txt
#	src/compiler/HipsyclClangPlugin.cpp

fodinabor committed:
Adapt sub_group functions and group_reduce to use RV intrinsics.

The RV intrinsics are only used if `HIPSYCL_HAS_RV` is defined. Then RV is required to run with `-mllvm -rv-lower` to support the intrinsics.

fodinabor committed:
Move work-item loop in separate function.

The compiler passes had to be adapted, as this might lead to the work-item loops not being the 2nd level loops anymore.

fodinabor committed:
Update tests to be (at least compilable) with RV.

fodinabor committed:
Add some preliminary sub_group size detection.

As the RV vector width is client code dependent and hipSYCL isn't necessarily compiled with the same -march=.., this has to be moved elsewhere at some point.

fodinabor committed:
Fix sub group reduce.

fodinabor committed:
Potentially improve group_reduce by sub_group reducing less often.

fodinabor committed:
Mark Loops with llvm.loop.vectorize.width if reqd_sub_group_size attribute present.

fodinabor committed:
Fix pocl pipeline with reqd_work_group_size.

fodinabor committed:
Fix WAR dependency in add_modulo.

fodinabor committed:
Only mark top level loops WI and remove noinline again (!O0).

fodinabor committed:
[CBS] Don't ever arrayify IndVar.

This ain't help too much, as usually the actually used index is offset and so on, thus still arrayified.

fodinabor committed:
[CBS] New mode: CBS without C++ work-item loop.

This allows testing with different marking different loops as parallel for multi-dimensional problems.
Still very much experimental.

fodinabor committed:
[LoopSplit] Add sycl_dgemm nd_range test.

Stolen from https://github.com/UoB-HPC/sycl_dgemm/blob/main/dgemm.cpp

fodinabor committed:
[CBS] Use a single pseudo IndVar.

This is required as all SubCFGs have to know the same IndVar to be replaced by the actual IndVar in their replica.

fodinabor committed:
[LoopSplit] Fully flatten kernels without WI Loop.

This is required for the new WILoop-less mode to get rid of additional allocas and so on.

fodinabor committed:
[LoopSplit] Add 3d stencil LIT test.

fodinabor committed:
[CBS] Don't erase IndVar if not manually created.

fodinabor committed:
[LoopSplit] Allow extra text in HIPSYCL_PIPELINE.

fodinabor committed:
[CBS] Remove viewCFG.

fodinabor committed:
[CBS] Fix nd_item regex.

fodinabor committed:
[CBS] Create loops around kernels without barrier.

fodinabor committed:
[CBS] Delete all loads from dummy vars and change Dim regex.

fodinabor committed:
[CBS] Only remove global variables if no uses left.

fodinabor committed:
[CBS] Only create WI loops, if none present.

fodinabor committed:
Add cbs-manual-loops to CI. (#4)

fodinabor committed:
Compat with LLVM 14+

llvm::PassBuilder::OptimizationLevel is renamed to llvm::OptimizationLevel.

fodinabor committed:
Import UniAna from RV.

fodinabor committed:
[CBS] Begin integrating UniAna in SubCFG.

But don't really use it, as uniform allocas would need at least double buffering and / or require not marking them as arrayified, s.t. they aren't reused when fixing up single subcfg values.

Note: won't support LLVM 11.

fodinabor committed:
[CBS] Use UniAna for contiguous values.

The idea behind this, is that e.g. the global ids in theory are contiguous values but currently are stored inside an arrayified alloca, which makes it impossible for e.g. the vectorizer to detect that loads to global memory are not random access but indeed contiguous.

To actually greatly benefit from this, we have to be able to identify the global id as contiguous. This does currently not happen, as the id.linear_id stuff leaves a loop at the stage we use it, with the global id as PHI node.
As the initial value stored to the ex_phi alloca is 0, it's first detected as uniform but the following values are contiguous. The VectorShape::join correctly identifies the alloca as varying then.

fodinabor committed:
[CBS] Only replace cont values, if not defined in SubCFG.

fodinabor committed:
Loop rotate loops in id and range.

fodinabor committed:
Replace rv namespace with hipsycl::compiler.

fodinabor committed:
[CBS] Fix CBS with omp loops.

fodinabor committed:
Import SyncDependenceAnalysis from LLVM main to support LLVM <= 11.

fodinabor committed:
[CBS] Allow keeping PHI's around.

fodinabor committed:
[UniAna] Replace getArgOperand with args / arg_size for LLVM 14+ compat.

fodinabor committed:
[CBS] Run SROA only for !O0, mark all local ids as cont, cleanups and  some todos added.

fodinabor committed:
[CBS] When using PHIs, we can actually use single element allocas for uniform values.

Note, the arrayification of already present allocas is not yet prevented.
Also single element allocas are no longer 64 byte aligned..

fodinabor committed:
Attempt enabling scalable vectorization if available.

This is mostly interesting for A64FX currently.

fodinabor committed:
[CBS] local_size parameter might become coerce 2, 3 element array as well.

In this case, extractvalue has to be used instead of the partial values / gep + load.

fodinabor committed:
Fix alignment of static local memory.

Still to fix is the alignment of the dynamic allocation.

fodinabor committed:
Fix contiguous index for 2D kernels without barrier.

fodinabor committed:
Mute UniformityAnalysis output.

fodinabor committed:
Demote PHIs on LLVM 12.

Otherwise also don't demote PHIs for old PM.

fodinabor committed:
Fix location of HIPSYCL_NO_PHIS_IN_SPLIT definition.

fodinabor committed:
Fix location of version Check in SDA.

fodinabor committed:
Only mark loops parallel if LLVM >= 12.0.1

fodinabor committed:
Create global var load if none found.

fodinabor committed:
Always use manually created loops.

Deactivate other pipelines in CI than CBS.

fodinabor committed:
[LoopSplit] Prune all but CBS pipelines.

fodinabor committed:
[LoopSplit][CI] Fix CBS CI pipeline.

-fpass-plugin was passed twice..

fodinabor committed:
[LoopSplit] Remove WILoop paths in SubCfgFormation and Barrier canonicalization.

fodinabor committed:
[LoopSplit] Remove superfluous WILoopMarker pass.

fodinabor committed:
[LoopSplit] Remove WILoop support for KernelFlattening, LoopSplitterInlining and PHIsToAllocas.

fodinabor committed:
[LoopSplit] Refactor LoopsParallelMarker pass.

fodinabor committed:
[LoopSplit] LoopSimplify weirdly is required, but at least make it FP.

fodinabor committed:
[LoopSplit] Make finding the local size value name independent.

This is necessary, so that -fno-discard-value-names doesn't have to be specified anymore.

fodinabor committed:
[LoopSplit] Toposort contiguous instructions before cloning.

This prevents use-before-def errors.

fodinabor committed:
[LoopSplit] Fix LLVM 11 compat.

fodinabor committed:
[LoopSplit] Add WITH_ACCELERATED_CPU and --hipsycl-use-accelerated-cpu options to CMake and syclcc.

WITH_ACCELERATED_CPU in the hipSYCL build is used to selectively build the plugin including the CBS pipeline or not. It also sets the default value for syclcc's `--hipsycl-use-accelerated-cpu`.

Also use HIPSYCL_USE_ACCELERATED_CPU instead of HIPSYCL_NO_FIBERS to enable compiler supported code sections of the nd-range code.

Client side build systems currently should either rely on the default value or explicitly specify `--hipsycl-use-accelerated-cpu` in the CXX_FLAGS.

fodinabor committed:
[LoopSplit] Prune usage of RV intrinsics.

fodinabor committed:
[LoopSplit] Hide uses of clang::annotate behind macros that are only active if clang is used.

fodinabor committed:
[LoopSplit] Move splitting related files to compiler/cbs.

fodinabor committed:
Get rid of debug output in test.

fodinabor committed:
Move subgroup property getter back to source file.

fodinabor committed:
[LoopSplit] Remove usage of deprecated LLVM functions.

This is mostly due to opaque ptr.

fodinabor committed:
Prune static_property_list again.

(Revert of merge 20e9e662 + removing changes depending on it thereafter)

fodinabor committed:
[LoopSplit] Add HIPSYCL_USE_ACCELERATED_CPU CMake option to client integration.

For this to be actually useful, syclcc-clang flags now can also have `=false/off/0` to explicitly disable the flag - to override the default syclcc.json value.

fodinabor committed:
Reactivate CI for Win and Mac + Linux CBS fixes

fodinabor committed:
Merge origin/develop into feature/prep-cbs-upstreaming.

fodinabor committed:
[LoopSplit] Merge fixup - must still check HIPSYCL_HAS_FIBERS.

fodinabor committed:
[LoopSplit] Disable CPU compiler support for NVC++.

fodinabor committed:
[LoopSplit] Update syclcc output in docs.

fodinabor committed:
[LoopSplit] Add some "user facing" documentation re cbs.

fodinabor committed:
[LoopSplit] Provide some in-code comments..

fodinabor committed:
[LoopSplit] Make ./sycl_tests fail CI again.

fodinabor committed:
[LoopSplit] Fix LLVM 14 compat (for now).

illuhad commented:
Can you have a glance at the `syclcc` changes in #683 and check if it might make sense to utilize some of the infrastructure from there here as well, e.g. backend flavors?

fodinabor commented:
I guess one could add an `omp.accelerated` flavor of the OpenMP target that mimics/replaces the `--hipsycl-use-accelerated-cpu` flag. The omp invocation could then return a caveat warning if the host compiler is not the known clang as well.

The latter is a slight inconvenience I hoped might be easily solvable by your changes, but from looking at them, doesn't seem that way. This is as even though the `config.pure_cpu_compiler` is checked in `omp_invocation` whether it is clang or not. If it's not clang, the `nd_range` acceleration is not enabled, even if `--hipsycl-use-accelerated-cpu` is set. If another backend (e.g. CUDA or HIP) is active though, the host compiler is overridden by clang, thus the acceleration would work. Now the question is: if we have an `omp.accelerated` flavor or so (and also `--hipsycl-use-accelerated-cpu` if we keep that flag), should that override the host compiler to clang or not? If not, can we _nicely_ add another stage, where depending of the finally chosen host compiler can add more flags? (or is that even already covered by something like `_configure_multipass_extended_host_pass`?)

illuhad commented:
> I guess one could add an omp.accelerated flavor of the OpenMP target that mimics/replaces the --hipsycl-use-accelerated-cpu flag. The omp invocation could then return a caveat warning if the host compiler is not the known clang as well.

I think this would nicely mimic the explicit multipass logic: The `--hipsycl-use-accelerated-cpu` flag would enable CPU acceleration for all (we currently only really have one) host backends, while the `.accelerated` would allow for a more selective enabling if multiple host backends are available.

> Now the question is: if we have an omp.accelerated flavor or so (and also --hipsycl-use-accelerated-cpu if we keep that flag), should that override the host compiler to clang or not

I think so - this would mimic the behavior of the device backends. The user has explicitly requested the flavor, so we should make sure that things happen as requested, no?

> If not, can we nicely add another stage, where depending of the finally chosen host compiler can add more flags? (or is that even already covered by something like _configure_multipass_extended_host_pass?)

The explicit multipass PR adds a `set_host_compiler()` member function for the multipass backends which is invoked by the main driver before compilation so that multipass backends can set caveats accordingly. We could extend this mechanism to all backends, not only multipass backends. Then they could take decisions based on the host compiler.

illuhad commented on CMakeLists.txt:
```diff
@@ -160,6 +172,9 @@ endif()
 cmake_policy(SET CMP0005 NEW)
 add_definitions(-DHIPSYCL_DEBUG_LEVEL=${HIPSYCL_DEBUG_LEVEL})
+if(WITH_ACCELERATED_CPU)
+  add_definitions(-DHIPSYCL_USE_ACCELERATED_CPU)
```
Does this need to be set globally or just for the clang plugin target?

fodinabor replied:
In this version globally, as the hw manager also needed to know this, but as I changed that, I also moved this.

illuhad commented on CMakeLists.txt:
```diff
@@ -284,11 +299,11 @@ if(NOT CUDA_CXX_FLAGS)
 
 if(NOT OMP_CXX_FLAGS) 
-  set(OMP_CXX_FLAGS "-I${Boost_INCLUDE_DIR} ${DEFAULT_OMP_FLAG}" CACHE STRING "Arguments passed to compiler to compile SYCL applications with OpenMP")
+  set(OMP_CXX_FLAGS "-I${Boost_INCLUDE_DIR} ${DEFAULT_OMP_FLAG} -D_ENABLE_EXTENDED_ALIGNED_STORAGE" CACHE STRING "Arguments passed to compiler to compile SYCL applications with OpenMP")
```
Why is `_ENABLE_EXTENDED_ALIGNED_STORAGE` required?

fodinabor replied:
It's actually Windows specific, but I figured it wouldn't hurt to keep it set generally while not cluttering the CMake file anymore..
It's a macro changing `std::aligned_storage`'s ABI to allow the necessary alignment (for `sycl::vec<double, 16>`)

illuhad replied:
Okay, that's fine. Maybe add a comment though?

illuhad commented on bin/syclcc-clang:
```diff
@@ -913,12 +930,15 @@ class omp_invocation:
-    flags =["-D__HIPSYCL_ENABLE_OMPHOST_TARGET__"]
+    flags = ["-D__HIPSYCL_ENABLE_OMPHOST_TARGET__"]
+    if self._use_plugin and not sys.platform.startswith("win32"):
+      flags += ["-fplugin=" + config.hipsycl_plugin_path, "-fpass-plugin=" + config.hipsycl_plugin_path, "-DHIPSYCL_USE_ACCELERATED_CPU"]
```
Maybe multiple lines? We also have the vague convention that internal macros set by `syclcc` to control compilation follow the form `__HIPSYCL_*__`. Would it make sense to apply this to `HIPSYCL_USE_ACCELERATED_CPU`?

fodinabor replied:
Will do. It might make sense to just relay the `-DWITH_ACCELERATED_CPU` CMake Option to a Macro for the hipSYCL runtime and compiler, while `-D__HIPSYCL_USE_ACCELERATED_CPU__` might be a better fit in syclcc and thus the client code.

illuhad commented on doc/compilation.md:
```diff
@@ -13,13 +13,32 @@ hipSYCL distinguishes two modes of compilation:
+
+## Compiler support to accelerate nd_range parallel_for on CPUs
+While hipSYCL usually strives to reuse as much compiler infrastructure as possible,
+the nd_range parallel_for is not efficiently implementable without explicit compiler support.
```
Maybe also add a sentence about scoped parallelism. I think the message should be:
1.) Use a model that is designed to be performance portable such as scoped,
2.) If this is not an option and you need nd_range (e.g. because of existing code base) the clang plugin is a viable option
3.) nd_range without clang plugin will be slow *as soon as barriers are present* due to issues with the `nd_range` model

illuhad commented on include/hipSYCL/compiler/.clang-format:
```diff
@@ -0,0 +1,2 @@
+BasedOnStyle: LLVM
+ColumnLimit: 120
```
Can we reduce this a bit? I think for the most part our lines are kept way shorter.

fodinabor replied:
Lowered to 100. Only touched new files (for fewer conflicts with #683 )

illuhad commented on include/hipSYCL/common/debug.hpp:
```diff
@@ -113,5 +119,14 @@ class output_stream {
+    content;\
+  }
 
+#define HIPSYCL_DEBUG_EXECUTE_VERBOSE(content) HIPSYCL_DEBUG_EXECUTE(HIPSYCL_DEBUG_LEVEL_VERBOSE, content)
```
Why do we need the `HIPSYCL_DEBUG_EXECUTE`? Isn't the same check effectively already don in `HIPSYCL_DEBUG_STREAM`?

fodinabor replied:
yeah, it is.. the difference here is that we can guard the execution of anything.. which sometimes came in handy, e.g. `HIPSYCL_DEBUG_EXECUTE_VERBOSE(F.viewCFG());` is used several times, which writes the functions CFG to a file and opens a text editor to view it.

illuhad commented on include/hipSYCL/compiler/Frontend.hpp:
```diff
@@ -438,6 +447,7 @@ class FrontendASTVisitor : public clang::RecursiveASTVisitor<FrontendASTVisitor>
   std::unordered_set<clang::FunctionDecl*> MarkedKernels;
   std::unordered_set<clang::FunctionDecl*> HierarchicalKernels;
   std::unordered_set<clang::FunctionDecl*> UserKernels;
+  std::unordered_set<clang::FunctionDecl*> NDKernels;
```
I wonder if there is a way to separate device backend compilation logic and CPU compilation logic somewhat.. Maybe prefix NDKernels with Host, or LoopFission?

illuhad commented on include/hipSYCL/compiler/cbs/IRUtils.hpp:
```diff
@@ -0,0 +1,158 @@
+static constexpr const char BarrierIntrinsicName[] = "__hipsycl_barrier";
+static constexpr const char LocalIdGlobalNameX[] = "_ZN7hipsycl4glue12omp_dispatch20__hipsycl_local_id_xE";
+static constexpr const char LocalIdGlobalNameY[] = "_ZN7hipsycl4glue12omp_dispatch20__hipsycl_local_id_yE";
+static constexpr const char LocalIdGlobalNameZ[] = "_ZN7hipsycl4glue12omp_dispatch20__hipsycl_local_id_zE";
```
Can we find a way to not hardcode things to OpenMP? It would be great if we could also just use this in say, a TBB backend

fodinabor replied:
Sure.. this is just a declaration-only global variable's mangled name... moving the variables to a different namespace (e.g. just `hipsycl::glue`) or marking them as `extern "C"` also removes the omp reference... alternatively we could again use `clang::annotate("...")` but I don't see a benefit there...

illuhad commented on include/hipSYCL/compiler/cbs/SubCfgFormation.hpp:
```diff
@@ -0,0 +1,70 @@
+#include "llvm/Pass.h"
+
+#if LLVM_VERSION_MAJOR <= 12
+#define HIPSYCL_NO_PHIS_IN_SPLIT
```
Can you add a comment about the meaning of this? Is there something unsupported in LLVM <= 12?

fodinabor replied:
I kind of believe that was just for making my CI test both cases... removed it..

illuhad commented on include/hipSYCL/compiler/cbs/SyncDependenceAnalysis.hpp:
```diff
@@ -0,0 +1,97 @@
+///
+/// This analysis relates points of divergent control to points of converging
+/// divergent control. The analysis requires all loops to be reducible.
+class SyncDependenceAnalysis {
```
I understand that this is from LLVM >= 13 that we backport for compatibility reasons, right? Would it make sense to put such classes in a `compatibility` or `llvm_pre13_compatibility` namespace to not clutter the main `hipsycl::compiler` namespace with content that is only needed for old LLVM versions?

illuhad commented on include/hipSYCL/compiler/cbs/UniformityAnalysis.hpp:
```diff
@@ -0,0 +1,162 @@
+//===- rv/analysis/VectorizationAnalysis.h - divergence analysis --*- C++ -*-===//
+//
+// Adapted from the RV Project, under the Apache License v2.0 with LLVM Exceptions.
```
This is from region vectorizer? Didn't you find that RV was still too unstable to rely on it?

fodinabor replied:
I pulled the uniformity analysis from RV into hipSYCL, yes.
The actual vectorization + RV's intrinsic were too unstable indeed, but with small adaptions from the original, the uniformity analysis has been pretty stable.

illuhad replied:
Okay. Are these small adaptions pointed out somewhere? It might be useful to know for maintenance in the future :)

illuhad commented on include/hipSYCL/glue/kernel_names.hpp:
```diff
@@ -92,17 +92,16 @@ struct kernel_name_traits<__hipsycl_unnamed_kernel, KernelBodyT> {
       multiversioned_kernel_wrapper<KernelBodyT, MultiversionParameters...>;
 
-  template <typename... MultiversionParameters>
+template <typename... MultiversionParameters>
```
This indentation looks a bit weird to me?

illuhad commented on include/hipSYCL/glue/omp/omp_kernel_launcher.hpp:
```diff
@@ -227,30 +259,58 @@ inline void parallel_for_kernel_offset(Function f,
+          }
+        }
+      }
+    }
```
This whole pattern is effectively what is already implemented in `iterate_range_omp`, no? Can we reuse it here?

illuhad commented on include/hipSYCL/runtime/omp/omp_hardware_manager.hpp:
```diff
@@ -30,8 +30,10 @@
+namespace hipsycl
+{
+namespace rt
+{
```
We usually don't do line breaks before namespace opening brace

illuhad commented on include/hipSYCL/sycl/libkernel/detail/local_memory_allocator.hpp:
```diff
@@ -159,7 +159,7 @@ class host_local_memory
   inline static char* _local_mem;
   
-  alignas(alignof(std::max_align_t)) inline static char _static_local_mem
+  alignas(sizeof(double) * 16) inline static char _static_local_mem
```
Why can we not have maximum alignment?

fodinabor replied:
`std::max_align_t` represents the alignment of `malloc` or similar, and is usually defined to `alignof(long double) == {8, 16}`. We need an alignment of `alignof(sycl::vec<double, 16>) == sizof(double) * 16`

illuhad replied:
Ugh. That's good to know. @DieGoldeneEnte I believe this might have an impact on the local memory used for the group algorithms on CPU.

illuhad commented on include/hipSYCL/sycl/libkernel/backend.hpp:
```diff
@@ -89,8 +89,14 @@
 #else
  #define HIPSYCL_FORCE_INLINE inline
+ #define HIPSYCL_ND_KERNEL
+ #define HIPSYCL_BARRIER
```
Should we maybe clarify those macros to refer to loop splitting? Like `HIPSYCL_LOOP_SPLIT_TRIGGER` or something similar?  After all, these attributes trigger something that is fairly specific to CPU backends and don't affect barriers or nd kernels *in general*.

illuhad commented on include/hipSYCL/sycl/libkernel/id.hpp:
```diff
@@ -110,14 +110,16 @@ struct id {
   id(const range<dimensions> &range) {
-    for(std::size_t i = 0; i < dimensions; ++i)
+    this->_data[0] = range[0];
+    for(std::size_t i = 1; i < dimensions; ++i)
```
This is curious, can you add a comment why this is beneficial?

fodinabor replied:
hmm.. I loop rotated all id and range functions as it completely eliminates the loop already at the point in time, the CBS happens, and also to fix some values being misclassified as varying by the uniformity analysis while they could be marked uniform. I don't think I verified that this is necessary for all these functions and possibly got a bit over-ambitious, but for most it was actually beneficial.

illuhad replied:
Ok, please add a comment (at least where it is really important) so that we don't accidentally "simplify" this in the future.

illuhad commented on src/compiler/GlobalsPruningPass.cpp:
```diff
@@ -29,7 +28,7 @@ void pruneUnusedGlobals(llvm::Module &M) {
   }
   HIPSYCL_DEBUG_INFO << "===> IR Processing: Pruning of globals complete, removed "
-                     << VariablesForPruning.size() << " global variable(s)." << std::endl;
+                     << VariablesForPruning.size() << " global variable(s).\n";
```
Why the changes in this file?

fodinabor replied:
I changed the plugin's `HIPSYCL_DEBUG_INFO` and friends to use `llvm::outs()` instead of `std::cout` to get more deterministically interleaved output from LLVM and the hipSYCL plugin. `llvm::outs()` does not support `<< std::endl` thus I had to replace it with `\n`.

illuhad replied:
Okay.

illuhad commented on src/compiler/HipsyclClangPlugin.cpp:
```diff
@@ -40,38 +45,77 @@ namespace hipsycl {
+#if defined(HIPSYCL_USE_ACCELERATED_CPU) && !defined(_WIN32)
+static llvm::RegisterPass<SplitterAnnotationAnalysisLegacy>
+    splitterAnnotationReg("splitter-annot-ana", "hipSYCL splitter annotation analysis pass",
+                          true /* Only looks at CFG */, true /* Analysis Pass */);
```
In PR #683, we relay information to the clang plugin using macros, i.e. `syclcc` will define a macro, and the clang plugin checks the existence of the macro as compiler option to adjust its behavior. Would it make sense to have a similar mechanism with the cbs plugin, so that we can disable support even though the clang plugin has been compiled with support for it?

fodinabor replied:
Sounds useful. It won't lead to functional changes, though, the plugin doesn't do anything if it doesn't find any nd_range annotated kernels, which only exist if `HIPSYCL_USE_ACCELERATED_CPU` (or its successor) is defined in the client code. Thus we could only reduce compiletime overhead slightly by checking the preprocessor definitions.
I'm not sure, if the right invocation context would be accessible from here either, so delaying this as possible optimization..

illuhad replied:
Alright - I think it's not super urgent anyway because, as you point out, it shouldn't make a functional difference. It might be useful though if CBS triggers issues with future clang versions etc.

illuhad commented on src/runtime/omp/omp_hardware_manager.cpp:
```diff
@@ -142,7 +142,11 @@ omp_hardware_context::get_property(device_uint_property prop) const {
     return std::numeric_limits<std::size_t>::max();
     break;
   case device_uint_property::max_group_size:
+#ifndef HIPSYCL_USE_ACCELERATED_CPU
```
I think this is confusing. There are two different ways in which this macro is used: At compile time of hipSYCL, affecting the hipSYCL runtime and clang plugin, and secondly, set by syclcc if accelerated cpu is enabled. What happens if `syclcc` compiles without this macro because the user has disabled accelerated cpu, but the runtime and clang plugin have it baked in? Would it make sense to distinguish one macro the runtime and clang plugin whether there is *support*, and a second one to be used by `syclcc` whether it is *enabled*?

fodinabor replied:
Split those macros to `HIPSYCL_WITH_ACCELERATED_CPU` and `__HIPSYCL_USE_ACCELERATED_CPU__`
Additionally, as discussed offline, changed this instance to default to `1024` as it also is more factually correct for the group algos.

illuhad commented on src/runtime/omp/omp_hardware_manager.cpp:
```diff
@@ -266,22 +270,21 @@ omp_hardware_context::get_property(device_uint_property prop) const {
+  }
+  assert(false && "Invalid device property");
+  std::terminate();
+}
```
This has just been moved down a couple of lines? :D

fodinabor replied:
yeah.. had it in the header for some period of time, which didn't change what I wanted to change though :)

illuhad commented on tests/sycl/group_functions/group_functions.hpp:
```diff
@@ -209,7 +209,7 @@ elementType<T> local_value(size_t id, size_t offsetBase) {
                                   size_t global_size) {
   BOOST_REQUIRE(global_size == 4 * local_size);
-  BOOST_REQUIRE(local_size + 10 < 2 * local_size);
+  BOOST_REQUIRE(local_size + 2 < 2 * local_size);
```
Reason for change?

illuhad commented on .github/workflows/linux-cbs.yml:
```diff
@@ -0,0 +1,70 @@
+          mkdir /home/runner/work/hipSYCL/hipSYCL/build/tests-cpu
+          cd /home/runner/work/hipSYCL/hipSYCL/build/tests-cpu
+          
+          cmake -DHIPSYCL_PLATFORM=cpu -DHIPSYCL_USE_ACCELERATED_CPU=true -DhipSYCL_DIR=/home/runner/work/hipSYCL/hipSYCL/build/install/lib/cmake/hipSYCL -DCMAKE_CXX_COMPILER=/usr/bin/clang++-${{matrix.clang_version}} -DCMAKE_CXX_FLAGS="$CXXFLAGS" /home/runner/work/hipSYCL/hipSYCL/tests
```
Plz use `HIPSYCL_TARGETS` for new code ;)

illuhad commented on .github/workflows/linux-cbs.yml:
```diff
@@ -0,0 +1,70 @@
+      - name: run CPU tests
+        run: |
+          cd /home/runner/work/hipSYCL/hipSYCL/build/tests-cpu
+          LD_LIBRARY_PATH=/home/runner/work/hipSYCL/hipSYCL/build/install/lib ./sycl_tests
```
This tests CPU exclusively, would there be value in also testing the use case when accelerated CPU and device backends are used simultanteously? (e.g. in hipSYCL-ci-bot?)

fodinabor replied:
I guess especially with the explicit multipass changes to syclcc it definitely would make sense to test at least the build...
But I'd delay that until explicit multipass and this are merged, so we can use the mix immediately.

illuhad reviewed (commented):
I'm so excited! :D

fodinabor committed:
Lower line length of .clang-format file in compiler sub dirs to 100.

fodinabor committed:
[LoopSplit] Small cleanups from review comments.

fodinabor committed:
[LoopSplit] Use iterate_range_omp_for for nd_group level.

fodinabor committed:
[LoopSplit] Split HIPSYCL_USE_ACCELERATED_CPU into two separate macros.

One HIPSYCL_WITH_ACCELERATED_CPU controls whether in the hipSYCL plugin, the CBS pipeline is inserted, the other __HIPSYCL_USE_ACCELERATED_CPU__ is the feature macro used at client compile-time to use the compiler based launcher instead of the fiber based launcher.

To avoid inconsistencies and also correctly align the max_group_size value with the already imposed limit of 1024 for the group functions, the CPU group size is now hardcoded to 1024.

Also prefix annotation macros with `HIPSYCL_LOOP_SPLIT_*`

fodinabor committed:
[LoopSplit] Update syclcc-clang to override host compiler to Clang, if --hipsycl-use-accelerated-cpu is set & update docs.

fodinabor committed:
[LoopSplit] Make __hipsycl_local_id_* `extern "C"` to get more general mangled names.

fodinabor committed:
[LoopSplit] Move SDA to pre_llvm12_compat namespace.

fodinabor reviewed (commented):
I changed syclcc to override the default host compiler, fi the accelerated CPU flag is set.
I will look into merging the multipass and this PR at some other point in time (possibly after one is merged ;-))

fodinabor committed:
[LoopSplit] Guard getting clang_path to fix MacOS.

fodinabor committed:
Basic documentation of the done adaptations.

Based on https://github.com/fodinabor/rv release/12.x.
The changes in that fork relative to upstram are subject to upstreaming.

fodinabor committed:
Add comments to explain macro and loop peeling.

fodinabor committed:
Propagate HIPSYCL_DEBUG_LEVEL through lit.

fodinabor committed:
Remove default setting of HIPSYCL_NO_PHIS_IN_SPLIT for LLVM < 13.

fodinabor committed:
Merge remote-tracking branch 'illuhad/develop' into feature/prep-cbs-upstreaming

fodinabor commented:
@hipsycl-ci-bot test hopefully one last time....

fodinabor committed:
Fix building Acc CPU + CUDA.

Currently quick and dirty handled by a static variable, but it should probably be prevented in syclcc that cpu_invocation and cuda_invocation both add the plugin flags.

fodinabor committed:
Merge remote-tracking branch 'origin/develop' into feature/prep-cbs-upstreaming

# Conflicts:
#	include/hipSYCL/compiler/Frontend.hpp
#	tests/sycl/sub_group.cpp

fodinabor committed:
[LoopSplit] Add omp.accelerated variant to syclcc.

illuhad commented on bin/syclcc-clang:
```diff
@@ -1193,7 +1236,16 @@ class compiler:
-        self._backends.append(omp_invocation(config))
+        if config.use_accelerated_cpu or \
+                (not "cuda-nvcxx" in config.targets
+                 and not sys.platform.startswith("win32") and not sys.platform.startswith("darwin")):
```
Should we not also check whether hipsycl-cpu-cxx is clang here?

fodinabor replied:
fixed.

fodinabor committed:
Revert "Fix building Acc CPU + CUDA."

This reverts commit b11c81738632da465c82522bf33c805683cae749.

fodinabor committed:
[LoopSplit] syclcc readying

- Actually remove duplicate -f(pass-)plugin from invocation.
- when "omp" is given, only default to accelerate if compiler is clang, too. For this normalize the clang and cxx paths using shutil.which, so that clang++ and /usr/bin/clang++ are the same.

fodinabor committed:
[LoopSplit] Allow empty clang_path.

fodinabor committed:
Merge remote-tracking branch 'illuhad/develop' into feature/prep-cbs-upstreaming

fodinabor committed:
[LoopSplit] Compat with LLVM<11

Dump plugin information to syclcc.json

With this information, the heuristic, whether "omp" shall be mapped to the accelerated variant doesn't stumble when a too old LLVM version is used.
Also use the information whether the cpu acceleration was built instead of the OS.

Also don't build the CPU acceleration support with too old clang.

fodinabor committed:
Update syclcc output in using-hipsycl.md.

Also add helper script to normalize the syclcc-clang --help output to be pasted into the docs.

illuhad commented on bin/syclcc-clang:
```diff
@@ -1125,6 +1174,59 @@ class omp_invocation:
+    return {
+      'requires-extended-host-pass' : False,
+      'extended-host-pass-providers' : [],
+      'conflicts' : ["cuda-nvcxx"],
```
```suggestion
      'conflicts' : ["cuda-nvcxx", "omp.library-only"],
```
Just noticed you can do `--hipsycl-targets="omp;omp.accelerated"` - this should probably be prevented, because we cannot actually have both simultaneously, right?

fodinabor replied:
good catch :)

fodinabor committed:
[LoopSplit] Add conflicts between omp.accelerated and omp.library-only.

illuhad commented:
When compiling the unit tests (clang 13, `--hipsycl-targets="omp.accelerated;cuda:sm_61;hip:gfx906"`) I see warnings such as:
```
[hipSYCL Warning] Instruction not dominated   %61 = lshr i64 %63, 1 operand:   %63 = phi i64 [ %61, %60 ]
```

This seems to happen for multiple source files. Is this expected?

Unit tests seem to run correctly however.

fodinabor committed:
[LoopSplit] Make domination warning an info, as it's being handled.

Also only print the extra info for the warning in loops parallel marker, only if warning level is active.

fodinabor commented:
Yeah, that's some relict I got too used to seeing :)
That's handled, so I just made it emit that only with debug level info, as it still might be relevant for actual debugging purposes..

illuhad commented:
Alright, in it goes <3
